documentation/projects/proposals/search_relevancy_sandbox/20230530-implementation_plan_staging_elasticsearch_reindex_dags.md

```diff
@@ -0,0 +1,231 @@
+Elasticsearch (ES) index creation ––full and proportional-by-provider–– which
+will allow us to decouple the process from the long Ingestion server's data
+refresh process and experiment with smaller indices. The DAG factories will
+create a DAG per media type supported, which currently consists `image` and
+`audio`. Also includes the adoption of two new index aliases for ease of
+handling and querying the new index types from the API with the
+[`internal__index`][api_ii_param] param.
+
+[api_ii_param]: https://github.com/WordPress/openverse/pull/2073
+
+## Expected Outcomes
+
+- DAGs for fully recreation of the index that the API uses which includes all
+  the database contents, by media type.
+- DAGs to create indexes of reduced size, proportional-to-production-by-provider
+  index, by media type.
+
+## Dependencies
+
+Same as for
+[Implementation Plan: Update Staging Database](/projects/proposals/search_relevancy_sandbox/20230406-implementation_plan_update_staging_database.md).
+
+This work is related to the [Staging database recreation
+DAG][staging_db_recreation] plan for having production volumes in the staging
+DB, which is expected to finish soon.
+
+[staging_db_recreation]: https://github.com/WordPress/openverse/issues/1989
+
+## DAGs
+
+### `recreate_full_<media_type>_staging_index` DAG
+
+#### Parameters
+
+1. `point_alias`: (Optional) A boolean value to indicate if the resulting index
+   will replace the current one pointed by the `<media_type>` alias. If `True`,
+   then the new index will be the one used by the staging API. Defaults to
+   `False`. Note: This is different from the "POINT_ALIAS" task of the ingestion
+   server, as the action performed in this DAG does not always imply the
+   deletion of the previous index and promotion.
+2. `delete_old_if_aliased`: (Optional) A boolean value to indicate if the old
+   index pointed by the `<media_type>` alias should be deleted after
+   replacement. Defaults to `False`.
+
+#### Outlined Steps
+
+This DAG will leverage the **Ingestion server's API** and use the existing
+[`REINDEX` task][reindex], which is the same used by the data refresh process to
+create the index.
+
+1. Get the current timestamp in order to create the index suffix with the form
+   of `full-<timestamp>`.
+2. Use the `ingestion_server.trigger_and_wait_for_task()` utility to send the
+   `REINDEX` call to ingestion server, passing the previously generated
+   `index_suffix` in the data payload.
+3. Once the index is created, make the alias `<media_type>-full` point to it.
+
+   1. Check if the alias exists. Use the
+      [`ElasticsearchPythonHook`][es_python_hook] with the
+      [indices.exists_alias][es_py_exists_alias] function.
+
+   2. If the alias doesn't exist, then it can be created and assigned in one
+      step using the [`indices.put_alias`][es_py_put_alias] function.
+
+   3. If the alias exists, send the request to add the new index and remove the
```

Comment:
In the case we have a previous index pointing to `<media_type>-full` that isn't aliased to `<media_type>` as well, should we also delete the old index once the new index is pointed to `<media_type>-full`? (Since the old index will essentially be dangling)?

Reply:
I was thinking of keeping all the indexes (unless specified otherwise) for benchmarking and testing purposes and deleting them manually. They could be switched manually in the API as well, but maybe I'm wrong in this assumption and we want to avoid dangling indexes as much as possible. What do you think? @AetherUnbound @sarayourfriend

Reply:
> Their use is totally optional. Should I highlight this more clearly in the proposal?

:+1: 

> [T]he proposal allows for multiple -full or -subset-by-provider indexes ... unless we keep telling the DAGs to delete the old ones

Just a heads up (I'm not entirely sure if this is the situation you're talking about), but if there are multiple indexes under the same alias, then ES will query _all_ those indexes with the alias, not just one. If `logs-2023` and `logs-2022` both have a `logs` alias applied to them, when querying with `logs` as the index name, both indexes will be queried and results from both will be included (not sure how scores are compared though). Not sure if that's the situation you were talking about by multiple indexes with the suffix aliases, or if you were talking about the aliases at all :sweat_smile: